### DISCUSSION TITLE
DDC issue on M2 Max

### DISCUSSION THREAD

**Discussion by Antieroe** [MonitorControl/MonitorControl#1298] [Q&A]
Hi everyone, I just replaced my macbook air M2 with the new Macbook pro M2 Pro, I use a LG ERGO Ultrafine 4k as monitor.
Previously I was able to change the monitor audio using the keyboard, now the audio goes up and down but without affecting the monitor.
<img width="467" alt="Screenshot 2023-01-24 alle 19 22 38" src="https://user-images.githubusercontent.com/123497662/214376191-68cf5f04-e55d-48ea-bea6-35ca4db29417.png">
<img width="820" alt="Screenshot 2023-01-24 alle 19 22 47" src="https://user-images.githubusercontent.com/123497662/214376202-ebfbe457-3b3f-4ab6-b353-7d0eb123e5ea.png">

**Comment by waydabber** [comment]
Wow, nice! However it seems like MonitorControl cannot detect your display as controllable via DDC. It might be due to architectural differences with the M2 Pro.

Can you try [BetterDisplay](https://betterdisplay.pro) if it works and can you show what do you see under Settings/Displays? That info can give us some more insight + we might start figuring out what's wrong.

<img width="942" alt="Screenshot 2023-01-24 at 21 14 19" src="https://user-images.githubusercontent.com/37590873/214398719-9bcd2bef-0c85-4e51-ad91-84efc7f7570b.png">

+ how is your display connected? Via USB-C or HDMI?

**Reply by Antieroe** [reply]
Hi, this is what I have
<img width="832" alt="Screenshot 2023-01-24 alle 22 18 21" src="https://user-images.githubusercontent.com/123497662/214418765-a8d83683-726f-4110-8d93-2a20e876fe33.png">
<img width="832" alt="Screenshot 2023-01-24 alle 22 18 37" src="https://user-images.githubusercontent.com/123497662/214418777-ff724e21-f339-4e22-a964-688bb2d88c0f.png">

**Reply by waydabber** [reply]
All right, this is not good. Is the display connected via USB-C or HDMI?

**Reply by Antieroe** [reply]
Usb-c

**Reply by waydabber** [reply]
All right. Clearly M2 MBPs work differently than M1 so neither BetterDisplay (and nor MonitorControl) can detect the right service port to do DDC control on the M2 Pro.

If you enter the following things in terminal, what is the outcome?

`ioreg -r -t -c DCPAVServiceProxy > DCPAVServiceProxy.txt`

`ioreg -r -t -c AppleCLCD2 > AppleCLCD2.txt`

`ioreg -r -t -n dp2hdmi-gpio > dp2hdmi.txt`

This should create 3 files. If you can attach them, I'll know a bit more.

**Reply by waydabber** [reply]
All right. Things are clearly super different. I'd need some more interaction. If you are ok with it tomorrow we can do a bunch of stuff.

Can you run an additional command so I can check how things look like in more depth and attach the result?

`ioreg -l >ioreg.txt`

**Reply by waydabber** [reply]
But on a warning note before you attach: this result should give a full layout of the system configuration and might contain serial numbers as well! If this is a problem, you can send it via other means if you are willing to help figuring this out. Thank you!

**Reply by waydabber** [reply]
Thank you! :) As a precaution I deleted the entry just to be on the safe side. I'll need some time analyzing the output (it's a bit late at night here so I'll continue it tomorrow), but will let me know my findings and if there is any version you could try out!

**Reply by waydabber** [reply]
Or, I'll just give it a shot and quickly post here a test version to see if that works as I might just nailed what's the difference. If it doesn't work, I'll continue looking.

Also, if you are willing and have a spare HDMI cable, can you try attaching your display via HDMI and send me over the same last file?

`ioreg -l >ioreg_hdmi.txt`

**Reply by waydabber** [reply]
Can you check if this BetterDisplay build works and can control the display via DDC over USB-C? Control over HDMI will most probably not work as that hardware is clearly different as well.

(download removed)

If this works, then I know what's wrong on M2 Pro and can implement things properly.

**Reply by Antieroe** [reply]
<img width="313" alt="Screenshot 2023-01-24 alle 23 49 10" src="https://user-images.githubusercontent.com/123497662/214438557-9f79dee4-a36c-41d0-bbbd-fdd90b102991.png">
it works but the keys on the keyboard don't really change the volume.
From the top panel instead you can change the audio of the monitor

**Reply by Antieroe** [reply]
in hdmi with the Mac "open" it seems to work

**Reply by Antieroe** [reply]
Now it also works with USB-C, I don't know why it didn't work before.
Now I have to go to sleep, tomorrow we can do more tests

**Reply by waydabber** [reply]
Hi there - it still might not work as from the brightness icon I see it's not using DDC. But it is because you have a HDR display running in HDR mode and BetterDisplay by default disables DDC when in HDR and switches to HDR upscale mode (that's why you can set a brightness beyond 100%). Can you try brightness control with HDR disabled?

Can you show me a screenshot of Settings/Displays?

Thank you!

**Reply by waydabber** [reply]
But if from the panel you can change the audio then it might work after all. If you want to the keyboard to change the audio, click on the `Set up Accessibility` under Tools and allow BetterDisplay in System Settings/Accessibility, then restart the app.

Let me know if things work (both with DP and HDMI, DDC brightness and volume).

If the app is using DDC mode to control the display, you should see a sun icon that is empty inside in the brightness slider. You can change the HDR setting in the app here:

<img width="280" alt="Screenshot 2023-01-25 at 6 50 20" src="https://user-images.githubusercontent.com/37590873/214490324-0c0724fd-7647-460f-b88a-9cc93f109792.png">

Under Settings/Displays you can enable DDC in HDR mode (if your is display is compatible with this), combined with HDR upscaling here:

<img width="942" alt="Screenshot 2023-01-25 at 6 52 30" src="https://user-images.githubusercontent.com/37590873/214490475-11fd049e-de53-4907-9757-2c1c61195403.png">

(when DDC is combined with HDR upscaling, the slider will be in combined dimming mode meaning it can both dim to zero, go beyond 100% and in the middle range DDC control is used instead of software dimming for added backlight control - this mode is signified with a full sun icon in the slider).

**Reply by Antieroe** [reply]
> Hi there - it still might not work as from the brightness icon I see it's not using DDC. But it is because you have a HDR display running in HDR mode and BetterDisplay by default disables DDC when in HDR and switches to HDR upscale mode (that's why you can set a brightness beyond 100%). Can you try brightness control with HDR disabled?
> 
> Can you show me a screenshot of Settings/Displays?
> 
> Thank you!

I've now turned HDR off
<img width="809" alt="Screenshot 2023-01-25 alle 14 40 03" src="https://user-images.githubusercontent.com/123497662/214578426-c93c9a9c-bb25-42c0-bb13-aaa2647aed38.png">

**Comment by Antieroe** [comment]
It seems that it takes a moment for the monitor to work but then it works without problems, will monitorcontrol also be updated?

**Reply by waydabber** [reply]
Yes, if this proves stable, I'll update MonitorControl with M2 compatibility as well.

So you see the empty sun symbol and if you move the brightness slider, the display responds correctly? If you check the display's OSD, you see the value change reflected in the Monitor OSD. Right? Does it work properly over HDMI as well?

(note: brightness control and stuff like these are free in BD as well so until MC is not updated, you can keep using it even if you don't ask for your code :)).

**Reply by waydabber** [reply]
I released an updated version that will hopefully work fine for M2 Macs: https://github.com/waydabber/BetterDisplay/releases

**Comment by waydabber** [comment]
Also: https://github.com/MonitorControl/MonitorControl/discussions/1299
